Assets/BossRoom/Scripts/Client/UI/Lobby/JoinLobbyUI.cs

```diff
@@ -0,0 +1,160 @@
+        {
+            if (!string.IsNullOrEmpty(newCode))
+                m_LocalLobbySelected = new LocalLobby.LobbyData(newCode.ToUpper());
+        }
+
+        public void OnJoinButtonPressed()
+        {
+            m_LobbyUIManager.JoinLobbyRequest(m_LocalLobbySelected);
+            m_LocalLobbySelected = default;
+        }
+
+        public void OnRefresh()
+        {
+            m_LobbyUIManager.QueryLobbiesRequest();
+        }
+
+        protected override void UpdateObserver(LobbyServiceData observed)
+        {
+            ///Check for new entries, We take CurrentLobbies as the source of truth
+            var previousKeys = new List<string>(m_LobbyButtons.Keys);
+
+            foreach (var codeLobby in observed.CurrentLobbies)
+            {
+                var lobbyCodeKey = codeLobby.Key;
+                var lobbyData = codeLobby.Value;
+
+                if (!m_LobbyButtons.ContainsKey(lobbyCodeKey))
+                {
+                    if (CanDisplay(lobbyData))
+                    {
+                        CreateLobbyPanel(lobbyCodeKey, lobbyData);
+                    }
+                }
+                else
+                {
+                    if (CanDisplay(lobbyData))
+                    {
+                        UpdateLobbyButton(lobbyCodeKey, lobbyData);
+                    }
+                    else
+                    {
+                        RemoveLobbyButton(lobbyData);
+                    }
+                }
+
+                previousKeys.Remove(lobbyCodeKey);
+            }
+
+            foreach (var key in previousKeys)
+            {
+                // Need to remove any lobbies from the list that no longer exist.
+                RemoveLobbyButton(m_LocalLobby[key]);
+            }
+        }
+
+        public void JoinMenuChangedVisibility(bool show)
+        {
+            if (show)
+            {
+                m_JoinCodeField.text = "";
+                OnRefresh();
+            }
+        }
+
+        public void OnQuickJoinClicked()
+        {
+            m_LobbyUIManager.QuickJoinRequest();
+        }
+
+        private bool CanDisplay(LocalLobby lobby)
+        {
+            return lobby.Data.State == LobbyState.Lobby && !lobby.Private;
+        }
+
+        /// <summary>
+        /// Instantiates UI element and initializes the observer with the LobbyData
+        /// </summary>
+        private void CreateLobbyPanel(string lobbyCode, LocalLobby lobby)
+        {
+            var lobbyPanel = m_UIFactory
+                .InstantiateActive(m_LobbyPanelPrototype.gameObject, m_LobbyPanelPrototype.transform.parent)
+                .GetComponent<LobbyPanelUI>();
+
+            lobbyPanel.BeginObserving(lobby);
+            lobbyPanel.OnClicked.AddListener(LobbyPanelSelected);
+            lobby.onDestroyed += RemoveLobbyButton; // Set up to clean itself
+
+            m_LobbyButtons.Add(lobbyCode, lobbyPanel);
+            m_LocalLobby.Add(lobbyCode, lobby);
```

Comment:
These two dictionaries are only ever modified at the same time.  Can we merge them into a single dictionary with an object/struct that contains both items?  Alternatively, just expose the LocalLobby as a property on the LobbyPanel?